### DISCUSSION TITLE
Prisma ORM Manifesto: Clarity and Collaboration

### DISCUSSION THREAD

**Discussion by wmadden** [prisma/prisma#25772] [Announcements]
Today we published [the Prisma ORM Manifesto](https://pris.ly/orm-manifesto) to mark the start of a new chapter for Prisma. As the project has grown, it’s become increasingly challenging to maintain the repo, stay connected with our community, and ship new features all at once. 

The manifesto outlines how we’re addressing these challenges, the changes we’re making, and our concrete next steps.

We’re opening this discussion to hear your feedback, answer your questions, and learn from your thoughts on the manifesto. Let’s work together to shape the future of Prisma.

**Comment by tonypapousek** [comment]
Thanks for having an open discussion!

With the Rust -> TypeScript conversion, do y'all have any concerns around performance?

**Reply by medz** [reply]
I'm more concerned if there aren't any engines available anymore for Prisma adaptations in third-party languages.

**Reply by wmadden** [reply]
Regarding third-party clients: https://github.com/prisma/prisma/discussions/25772#discussioncomment-11445662

**Reply by aqrln** [reply]
Good question! First of all, this is a big transition, and it's not going to happen at once: we are committed to maintain stability and full functionality of the ORM, and the changes will be gradual. We will have a WebAssembly Rust component at the time being, however it will not interact with the database.

As for the performance, the biggest bottleneck right now is the serialization of data and passing it through the FFI boundary. This will go away, and the preliminary benchmarks on the prototype show massive performance improvements. More than that, the new architecture allows to prepare Prisma queries and re-use the prepared queries with virtually no runtime cost compared to using a raw SQL query.

We will share more details with the community soon, so stay tuned!

**Reply by tonypapousek** [reply]
@aqrln - Awesome, thanks for the detail!

I'm not too familiar with the code; does the `generate` command use Rust or TS currently?

**Reply by aqrln** [reply]
Both. It already doesn't use any natively compiled engines though, only the [`@prisma/prisma-schema-wasm`](https://www.npmjs.com/package/@prisma/prisma-schema-wasm) package. When writing a custom generator, you might not even be aware that any Rust code is used for schema parsing and generating the DMMF, it's pretty well encapsulated.

**Reply by aqrln** [reply]
That WASM module is fairly general purpose though and powers other things as well (like the language server and the VS Code extension). The `generate` command itself and the Prisma Client generator are all pure TypeScript.

**Comment by medz** [comment]
If the engine switches from Rust to TypeScript, does that mean there will be no more third-party language adaptations?

If so, this will be a disaster for Prisma Rust（ @Brendonovich ）, Prisma Go（ @steebchen ）, Prisma Python（ @RobertCraigie ）, and Prisma Dart （ @medz ）.

**Reply by steebchen** [reply]
wondering about this as well

**Reply by wmadden** [reply]
That's an excellent question and we're hearing it a lot! We hope to continue to enable third-party language clients but the precise path forward isn't clear yet.  Here's what we're thinking:

* We'd like to talk to third-party client maintainers to understand what parts of the existing Prisma architecture they depend on specifically
* In the next three months we'll release an early access preview of our new TS-based architecture to help guide the discussion on supporting them 
* If possible, we'll extract the Rust elements they depend on into a separate package maintained externally
* We want to make the code generation process extensible and adaptable to support generating assets besides the Prisma TS client

**Reply by medz** [reply]
> That's an excellent question and we're hearing it a lot! We hope to continue to enable third-party language clients but the precise path forward isn't clear yet. Here's what we're thinking:
> 
> * We'd like to talk to third-party client maintainers to understand what parts of the existing Prisma architecture they depend on specifically
> * In the next three months we'll release an early access preview of our new TS-based architecture to help guide the discussion on supporting them
> * If possible, we'll extract the Rust elements they depend on into a separate package maintained externally
> * We want to make the code generation process extensible and adaptable to support generating assets besides the Prisma TS client

To be honest, if the subsequent Prisma engine still builds WASM modules, I hope to provide CAPI versions simultaneously. Building WASM modules means it is easy to create CAPI copies.

The reason why I want to have a CAPI version so much is that it can provide engine support for any third-party language.

At present, I have paused the work on Prisma Dart and waited for Prisma v7. Starting from Prisma 6.2, Prisma can no longer support running on Flutter (capi engine is broken). But if I continue to use the binary version, it is not in line with the Prisma manifesto. So I paused.

**Comment by lukadriel7** [comment]
Thanks for the heads up. To be honest, I am a bit worried about the change. I have personally been following https://github.com/prisma/prisma-engines/pull/4208 because the integration of Geospatial features didn’t seem to be a priority for the core Prisma team. One year later, as I hoped it would eventually be integrated, the change to typescript is announced. My guess is the pull request is now meaningless and may be closed... The manifesto mentioning extensibility by design, I hope it will simplify the integration of new features. 

An aspect I am hopeful is the feature parity between the supported databases when possible. Having to run a Postgres instance just because SQLite provider doesn’t support JSON can be annoying.

**Reply by aqrln** [reply]
The pull request is not meaningless and can still be merged. There's a very long way to the full TypeScript core, and the code this PR changes is still going to be used in 2025, albeit in a different manner, with better architecture and without native binaries.

**Comment by kibertoad** [comment]
@wmadden As a company that primarily chose Prisma due to a strong CockroachDB support, should we be concerned that the support might be dropped along the way of the engine rewrite from Rust to TS? Or plan is to preserve feature parity, and only deprioritize further enhancements for non-core databases?

**Reply by wmadden** [reply]
Hi @kibertoad, you’re correct that CockroachDB won’t be a primary focus for us going forward, but we want to reassure you that support for CockroachDB isn’t going away. We’ll continue to prevent regressions, though some issues or new functionality may require community contributions since it’s not our primary focus.

That said, we’re actively working on making Prisma’s codebase more modular, which should make it easier for the community to contribute and extend support for databases like CockroachDB. For partners with an enterprise support contract, we can also allocate engineering resources to address specific issues.

**Reply by kibertoad** [reply]
thank you, this is good news!

**Comment by hinogi** [comment]
still waiting for that announcement from 2019 to enable ppl to create their own connectors to the database they want, to materialize.

**Reply by gniting** [reply]
@hinogi which db(s) are you looking to build a connector for?

**Reply by hinogi** [reply]
me, none because I am not a rust programmer. But I wanted support for oracle. like some others. for 5 years now. https://github.com/prisma/prisma/issues/2853 has been closed because it is not profitable.

**Reply by gniting** [reply]
> me, none because I am not a rust programmer. But I wanted support for oracle. like some others. for 5 years now. #2853 has been closed because it is not profitable.

Thanks for clarifying. You may not be a programmer, but you are driven to get this particular item done and I admire that kind of conviction! 🙇

Oracle support is not an easy lift (as one would imagine). The ORM is open source and will remain open source so this is not about profitability. It's about resources, and having those resources deliver what our community deems as most important (via upvotes/comments) and alignment with our goals as a team/company.

As the manifesto suggests, we're working down the list of most voted/commented items and when this one hits, we'll (have to) pick it up or clearly communicate that are not doing it. To accelerate that, let me take an action item on this. Expect to hear back from me on this topic in a week once I've had the chance to talk to the relevant team members internally.

**Reply by hinogi** [reply]
As you can derive from this very long issue. A lot of ppl offered their time and expertise to get it done, I assume that they cannot do it without getting their hands on crates from prisma. I further assume that those are not open source. So the resource argument is a strawman because other ppl would have done the job. Alignment to the goals comes also back to profitability or widespread support is not a team/company goal.
I don't like to be gaslit on a subject that started in 2020 and was neglected ever since.
I see it more straightforward, it can be done, a lot of ppl already offered to have it done, the only showstopper is prisma in this matter to enable them to make it happen. Prisma explicitly decided against it.
There was a reaction from a contributor after more like 1 year then after 3 years another and finally got shot down in last december.
The original oracle connector issue is even older https://github.com/prisma/prisma1/issues/1644 2018
https://github.com/prisma/prisma/discussions/15714 also got this from a maintainer who tried the same gaslighting on me.
I really not appreciated such kind of behavior.

**Reply by gniting** [reply]
>A lot of ppl offered their time and expertise to get it done, I assume that they cannot do it without getting their hands on crates from prisma.  I further assume that those are not open source. So the resource argument is a strawman because other ppl would have done the job. 

This is incorrect. The *entire* Prisma codebase is open source. If someone wanted to build an Oracle connector, they could do so and open a PR. There are already several examples of community-built driver adapters, which you can find here: https://www.prisma.io/docs/orm/overview/databases/database-drivers#community-maintained-database-driver-adapters.

> Alignment to the goals comes also back to profitability or widespread support is not a team/company goal.

You continue to assume this when I have clarified that profitability is not the case. The ORM is open source. We have an entirely separate team working on the ORM vs one that works on on commercial products. 

> The original oracle connector issue is even older https://github.com/prisma/prisma1/issues/1644 2018

That issue is from Prisma 1, which is not even the same package that we are talking about now, or the Prisma ORM as we currently know it.

Since then, we’ve introduced DriverAdapters, a feature that allows anyone to extend the ORM independently. Before this release, extending the ORM required significant collaboration with our team. With DriverAdapters, that dependency has been eliminated. We still carry the burden to merge PRs, but that is entirely doable should someone send a PR our way. We merge relevant PRs all the time.

> I don't like to be gaslit on a subject that started in 2020 and was neglected ever since.

"Neglect" is not the appropriate term. If this topic were neglected, we wouldn’t respond every time it is raised. Why do we engage? Because we are committed to addressing misconceptions. The manifesto, linked above, clearly outlines the databases that are our priority. These priorities are informed by research, community feedback/wants/asks, industry trends, and our strategic goals.

Our decisions come down to resource allocation and prioritization. While you might see this as a single unresolved issue, we view it in the context of all community needs. Allocating resources to something as complex as Oracle support could prevent us from addressing other, higher-priority community requests. We aim to avoid frustration from multiple angles by taking a more balanced approach.

>I see it more straightforward, it can be done, a lot of ppl already offered to have it done, the only showstopper is prisma in this matter to enable them to make it happen. Prisma explicitly decided against it.

It is never as straightforward as it seems. Supporting Oracle is a challenging and resource-heavy endeavor. Could we use community help, of course. But our code base is also not that easy to grok, and we recognize that. We are taking steps to address that as well. The recent announcement of decoupling of the Rust engine will certainly help with that.

> https://github.com/prisma/prisma/discussions/15714 also got this from a maintainer who tried the same gaslighting on me. I really not appreciated such kind of behavior.

The response you received was a direct reflection of our thinking at that time. Just like I am now outlining our thinking at the present moment. Communication is always subject to interpretation, but I can assure you that the Prisma team values constructive feedback and community collaboration. While our team strives to serve as many community members as possible, it is impossible to meet every request. That is an unfortunate reality, but we remain committed to transparency and respectful dialogue.

As I have mentioned in the comment above, give me a few days on this and I will circle back.

**Reply by hinogi** [reply]
Here it goes again. Driver Adapters are not the same as providers or connectors as far as I understand it. Hence it is not possible to just write an oracle adapter if you don't have a fitting provider written in **Rust**. If that is open source too, that would be great.
Yes, the issue is from prisma 1to underline that this issue exists since 2018.

**Reply by gniting** [reply]
> Driver Adapters are not the same as providers or connectors as far as I understand it

That understanding is incorrect. DriverAdapters serve as a foundational step toward writing exactly that: database driver adapters.

> Hence it is not possible to just write an oracle adapter if you don't have a fitting provider written in Rust

This highlights why this area is a moving target—our efforts are focused on removing the dependency on the Rust engine altogether and to make it even more easier for the community to contribute, all the while improving performance and reliability.

> If that is open source too, that would be great.

It is, as are all the other database drivers. If someone really wanted to take on writing an Oracle driver, the source material for everything is available to them:
* https://github.com/prisma/prisma-engines
* https://github.com/orgs/prisma/repositories?q=rust

> Yes, the issue is from prisma 1to underline that this issue exists since 2018.

That point is not directly relevant to the current discussion. Software evolves, and so should the framing of the discussion. Referencing the age of an issue does not inherently strengthen your argument, especially if the underlying architecture has changed significantly since then.

> Here it goes again.

I’d like to offer some constructive feedback. Comments like this can come across as passive-aggressive. I am engaged in a professional dialogue and have already communicated that I intend to discuss this matter with the team. I would appreciate it if you could refrain from such remarks. They risk undermining the professionalism of this exchange and may inadvertently reflect the same behavior you have criticized—revisiting a topic in a manner that could be perceived as gaslighting.

**Reply by gniting** [reply]
Following up on this...One of the key challenges in enabling external contributions for building an Oracle adapter has been the inherent complexity of our codebase. We recognize that this complexity has posed barriers, and our recent announcement regarding the removal of the Rust component(s) is aimed at reducing this challenge.

Dedicating resources to building something that we know will need to fundamentally change is not ideal. Therefore, for the near term our focus will be on delivering the necessary capabilities that pave the way towards delivering a code base that is easier and faster to work and create with.

Once that is out there, we'll revisit all requested enhancements, including an Oracle driver, and build out a roadmap.

**Comment by tonypapousek** [comment]
Howdy - has the ["splitting the `index.d.ts`"](https://github.com/prisma/prisma/issues/4807) problem been given any priority for 2025?

It causes some frustration when using Prisma at scale. Our editors get weird and refuse to interpret `d.ts` files after a size limit is hit, which prevents us from following types, etc. Bit painful when debugging TypeScript code.

Thanks again, I'm excited to see how Prisma grows over the next year!

**Reply by wmadden** [reply]
Hey @tonypapousek, this is definitely an issue we want to resolve in 2025 👍🏻

**Reply by tonypapousek** [reply]
> Hey @tonypapousek, this is definitely an issue we want to resolve in 2025 👍🏻 

That's excellent news, thanks!

**Comment by wnm210** [comment]
Well, I guess the change from Rust to TypeScript just killed all the major community maintained clients :/

https://github.com/steebchen/prisma-client-go/issues/1542
https://github.com/Brendonovich/prisma-client-rust/discussions/476

Been using the Go client [*]

**Reply by steebchen** [reply]
yeah pretty much, unfortunately.

**Reply by medz** [reply]
I am also ready to give up prisma-dart; because this move makes it impossible for me to continue to maintain prisma-dart.

It even forced me to switch the server built in Dart to Bun/Node for development. It is very heavy for me because it is difficult to find an alternative in Dart (there is not even a stable database connector under Dart)

**Comment by tyteen4a03** [comment]
This is extremely disappointing news coming from somebody who architectured a project based on the fact that Prisma *is* the perfect multi-language schema DSL.

The only edge for Prisma is now lost. I might as well go back to creating REST APIs and use Drizzle.

**Reply by gniting** [reply]
Thanks for sharing your feedback.
